demos/coreHTTP/http_demo_s3_download_multithreaded.c

```diff
@@ -0,0 +1,965 @@
+
+/**
+ * @brief Represents the network context used for the TLS session with the
+ * server.
+ */
+static NetworkContext_t xNetworkContext;
+
+/**
+ * @brief The host address string extracted from the pre-signed URL.
+ *
+ * @note httpexampleS3_PRESIGNED_GET_URL_LENGTH is set as the array length here
+ * as the length of the host name string cannot exceed this value.
+ */
+static char cServerHost[ httpexampleS3_PRESIGNED_GET_URL_LENGTH ];
+
+/**
+ * @brief The length of the host address found in the pre-signed URL.
+ */
+static size_t xServerHostLength;
+
+/**
+ * @brief Data type for request queue.
+ *
+ * In addition to sending #HTTPRequestHeaders_t, we need to send the buffer it
+ * uses as the pointer to the buffer in #HTTPRequestHeaders_t may not be
+ * accurate after being copied into the queue.
+ */
+typedef struct RequestItem
+{
+    HTTPRequestHeaders_t xRequestHeaders;
+    uint8_t ucHeaderBuffer[ democonfigUSER_BUFFER_LENGTH ];
+} RequestItem_t;
+
+/**
+ * @brief Data type for response queue.
+ *
+ * In addition to sending the #HTTPResponse_t, we need to send the buffer it
+ * uses as the pointer to the buffer in #HTTPRequestHeaders_t may not be
+ * accurate after being copied into the queue.
+ */
+typedef struct ResponseItem
+{
+    HTTPResponse_t xResponse;
+    uint8_t ucResponseBuffer[ democonfigUSER_BUFFER_LENGTH ];
+} ResponseItem_t;
+
+/*
+ * @brief Struct used for sending requests to the HTTP thread.
+ *
+ * This is used by both the main and HTTP threads. The global scope ensures that
+ * it will be located at the same address in the main and HTTP threads, so that
+ * pointers remain valid when copied over.
+ */
+static RequestItem_t xRequestItem = { 0 };
+
+/**
+ * @brief Struct used for receiving responses from the HTTP thread.
+ *
+ * This is used by both the main and HTTP threads. The global scope ensures that
+ * it will be located at the same address in the main and HTTP threads, so that
+ * pointers remain valid when copied over.
+ */
+static ResponseItem_t xResponseItem = { 0 };
+
+/**
+ * @brief Queue for HTTP requests. Requests are written by the main thread, and
+ * executed by the HTTP thread.
+ */
+static QueueHandle_t xRequestQueue;
+
+/**
+ * @brief Queue for HTTP responses. Responses are written by the HTTP thread,
+ * and read by the main thread.
+ */
+static QueueHandle_t xResponseQueue;
+
+/**
+ * @brief Handle of prvAsyncPublishTask.
+ */
+static TaskHandle_t xHTTPTask;
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Connect to HTTP server with reconnection retries.
+ *
+ * @param[out] pxNetworkContext The output parameter to return the created
+ * network context.
+ *
+ * @return pdFAIL on failure; pdPASS on successful connection.
+ */
+static BaseType_t prvConnectToServer( NetworkContext_t * pxNetworkContext );
+
+/**
+ * @brief Send an HTTP request based on a specified method and path, and
+ * print the response received from the server.
```

Comment:
Is this the correct description of this function? The description makes it sound like a generic HTTP request function, but the name mentions about S3 Download.

Reply:
No, thanks for catching that. Updated the comment.